graph-partitioning/README.rst

```diff
@@ -0,0 +1,82 @@
+
+One way to solve this problem is with the graph partitioning problem.  If we think of our group of people as a graph (a node/vertex for each person), and draw an edge between people that are friends, we have a model of our network.  If we look for a minimum cut in our graph, then we are looking for a way to split the nodes into two groups so that there are as few edges as possible between the groups.  In our group of peoople, this means that we have two teams with as few friends as possible on different teams.
+
+We will run the graph partitioning problem on an Erdos-Renyi random graph with probability 0.2 on 40 nodes to find the best way of splitting the graph into two subsets to minimize the number of cut edges.
+
+Usage
+-----
+To run the demo, type:
+::
+  python graph_partitioning.py
+
+After running, output will be printed to the command line that states whether or not the best (lowest energy) solution found is a valid partition of two equal-sized sets.  If the solution is valid (feasible), then the number of cut edges for that solution is displayed.
+
+Code Overview
+-------------
+The code implements a QUBO formulation of this problem.
+
+The answer that we are looking for is a partition of the nodes in the graph, so we will assign a binary variable for each node, i.e. variable x_i denotes whether node i is in one subset (call it Subset 0) or the other (Subset 1).
+
+The objective function that we are looking to optimize is minimizing the number of cut edges.  To count how many cut edges we have given a partition of the nodes (assignment of our binary variables), we consider a single edge in a graph in the table below.  We only want to count an edge if the endpoints are in different subsets, and so we assign a 1 for the edge column in this case and a 0 otherwise.
+
+=== === ==========
+x_i x_j edge (i,j)
+--- --- ----------
+0   0   0
+0   1   1 
+1   0   1
+1   1   0
+=== === ==========
+
+From this table, we see that we can use the expression x_i+x_j-2x_ix_j to calculate the edge column in our table.  Now for our entire graph, our objective function can be written as shown below, where the sum is over all edges in the graph.
+
+.. image:: readme_imgs/QUBO.png
+
+Next we need to consider our constraint:  Subset 0 and Subset 1 must have the same sizes.  We can measure the size of Subset 1 by summing up our binary variables.  To ensure the two subsets have the same size, we enforce a constraint that Subset 1 has size equal to half of all nodes in the graph.  We first consider how to represent this constraint mathematically using our chosen binary variables, and use the following equality to represent our constraint.
+
+.. image:: readme_imgs/constraint_1.png
+
+For a QUBO, we need our constraints to be represented by mathematical expressions that are satisfied at the minimum value.  For this constraint, we can use the following expression that has a minimum value of 0 that occurs when Subset 1 has size exactly `|V|`/2.
+
```

Comment:
I think this is actually true for the objective function, not necessarily the QUBO, because we're minimizing. Also "at the minimum value" opens the question "value of what?"

Reply:
Our constraints need to be converted from the equality form to the quadratic expression (think line converted to parabola).  The smallest/minimum value of that parabola will be the location where our constraint is satisfied.  We don't know what the value will be - it depends on how you write your expression.  Since it's an argmin, the value itself doesn't matter.

The objective function is the optimization part.  In this case, minimizing the number of cross edges.